File: dp/coin_change.py
"""Coin Change.

Min coins to make amount, or -1.

Time: O(amount * |coins|)
Space: O(amount)
"""


def coin_change(coins: list[int], amount: int) -> int:
    dp = [amount + 1] * (amount + 1)
    dp[0] = 0
    for a in range(1, amount + 1):
        for c in coins:
            if a - c >= 0:
                dp[a] = min(dp[a], dp[a - c] + 1)
    return dp[amount] if dp[amount] != amount + 1 else -1


if __name__ == "__main__":
    assert coin_change([1, 2, 5], 11) == 3
    assert coin_change([2], 3) == -1
    print("OK")

File: java/CoinChange.java
/*
 * Coin Change.
 *
 * Min coins to make amount, or -1.
 *
 * Time: O(amount * |coins|)
 * Space: O(amount)
 */

import java.util.Arrays;

public class CoinChange {

    public static int coinChange(int[] coins, int amount) {
        if (amount < 0) {
            return -1;
        }
        int sentinel = amount + 1;
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, sentinel);
        dp[0] = 0;
        for (int a = 1; a <= amount; a++) {
            for (int c : coins) {
                if (a - c >= 0 && dp[a - c] != sentinel) {
                    dp[a] = Math.min(dp[a], dp[a - c] + 1);
                }
            }
        }
        return dp[amount] == sentinel ? -1 : dp[amount];
    }

    public static void main(String[] args) {
        assert coinChange(new int[] {1, 2, 5}, 11) == 3;
        assert coinChange(new int[] {2}, 3) == -1;
        assert coinChange(new int[] {1}, 0) == 0;
        System.out.println("OK");
    }
}
